stl/inc/span

```diff
@@ -0,0 +1,678 @@
+            "Mismatch between static extent and size of initializing data");
+#if _CONTAINER_DEBUG_LEVEL > 0
+        if constexpr (_OtherExtent == dynamic_extent) {
+            _STL_VERIFY(_Ext._Size() == _Extent, "Mismatch between static extent and size of initializing data");
+        }
+#endif // _CONTAINER_DEBUG_LEVEL > 0
+    }
+
+    _NODISCARD constexpr size_t size() const noexcept {
+        return _Extent;
+    }
+};
+
+template <>
+struct _Span_extent_type<dynamic_extent> {
+    constexpr _Span_extent_type() noexcept = default;
+
+    constexpr explicit _Span_extent_type(const size_t _Size) noexcept : _Mysize(_Size) {}
+
+    template <size_t _OtherExtent>
+    constexpr explicit _Span_extent_type(const _Span_extent_type<_OtherExtent> _Ext) noexcept : _Mysize(_Ext._Size()) {}
+
+    _NODISCARD constexpr size_t size() const noexcept {
+        return _Mysize;
+    }
+
+private:
+    size_t _Mysize{0};
+};
+
+// STRUCT TEMPLATE _Is_span
+template <class _Ty, size_t _Extent>
+class span;
+
+template <class>
+struct _Is_span : false_type {};
+
+template <class _Ty, size_t _Extent>
+struct _Is_span<span<_Ty, _Extent>> : true_type {};
+
+// STRUCT TEMPLATE _Is_std_array
+template <class>
+struct _Is_std_array : false_type {};
+
+template <class _Ty, size_t _Extent>
+struct _Is_std_array<array<_Ty, _Extent>> : true_type {};
+
+// STRUCT TEMPLATE _Is_span_convertible_range
+template <class _Cont, class _Ty>
+struct _Is_span_convertible_range
+    : is_convertible<remove_pointer_t<decltype(_STD data(_STD declval<_Cont&>()))> (*)[], _Ty (*)[]> {};
+
+// STRUCT TEMPLATE _Has_container_interface
+template <class, class = void>
+struct _Has_container_interface : false_type {};
+
+template <class _Cont>
+struct _Has_container_interface<_Cont,
+    void_t<decltype(_STD data(_STD declval<_Cont&>())), decltype(_STD size(_STD declval<_Cont&>()))>> : true_type {};
+
+// STRUCT TEMPLATE _Is_span_compatible_container
+// clang-format off
+template <class _Cont, class _Ty>
+struct _Is_span_compatible_container : conjunction<
+	negation<is_array<_Cont>>,
+	negation<_Is_span<remove_const_t<_Cont>>>,
+	negation<_Is_std_array<remove_const_t<_Cont>>>,
+	_Has_container_interface<_Cont>,
+	_Is_span_convertible_range<_Cont, _Ty>> {};
+// clang-format on
+
+// [views.span]
+// CLASS TEMPLATE span
+template <class _Ty, size_t _Extent = dynamic_extent>
+class span : public _Span_extent_type<_Extent> {
+public:
+    using _Mybase                = _Span_extent_type<_Extent>;
+    using element_type           = _Ty;
+    using value_type             = remove_cv_t<_Ty>;
+    using index_type             = size_t;
+    using difference_type        = ptrdiff_t;
+    using pointer                = _Ty*;
+    using const_pointer          = const _Ty*;
+    using reference              = _Ty&;
+    using const_reference        = const _Ty&;
+    using iterator               = _Span_iterator<_Ty>;
+    using const_iterator         = _Span_iterator<const _Ty>;

```

Comment:
Is `iterator` convertible to `const_iterator`, and comparable/subtractable/etc? I mentioned this in an earlier comment, but I don't see the resolution here. If you need help getting the conversion right, I can show you an example. (The existing examples where `iterator` actually inherits from `const_iterator` are undesirable.)

Reply:
I plan on tackling that this week. With the previous _Span_iterator version I had some problems with comparability and decided to postpone them after the rewrite.

Shockingly I forgot and in my iterator tests the comparability is commented out ...

Reply:
So just to ensure that my thinking is right. `iterator` should be convertible to `const_iterator` but *not* vice versa.

If I understand it correctly there are 2 ways. I could create a constructor that implicitly converts `_Span_iterator` if the underlying pointer type is convertible

```cpp
template <class _OtherType, enable_if_t<is_convertible_v<_OtherType*, pointer>, int> = 0>
constexpr _Span_iterator(const _Span_iterator<_OtherType> _Other)
#if _ITERATOR_DEBUG_LEVEL >= 1
	: _Myptr(_Other._Myptr), _Mybegin(_Other._Mybegin), _Myend(_Other._Myend) {}
#else // ^^^ _ITERATOR_DEBUG_LEVEL >= 1 ^^^ // vvv _ITERATOR_DEBUG_LEVEL == 0 vvv
	: _Myptr(_Other._Myptr) {}
#endif // _ITERATOR_DEBUG_LEVEL
```
That should take care of the necessary conversions. I am unsure whether I would have to do any shenanigans removing pointer etc?!

On the other hand I could overload all the required operators.
```cpp
template <class _Ty1, class _Ty2>
_NODISCARD constexpr bool operator==(const _Span_iterator<_Ty>&, const _Span_iterator<_Ty2>&) noexcept;
```

I would greatly prefer the first version. However I am unsure whether that goes well with @CaseyCarters goal that `_Span_iterator` should be an aggregate?!

Reply:
Also do we need the non-member friend overloads of those oprators anyway?

Reply:
You definitely need the constructor (and yes, it is one-way; constructing `iterator` from `const_iterator` would allow users to remove constness which would be bad). Constraining it seems reasonable since users can overload functions on `span<X>::iterator` and expect the "right one" to be called. For the operators, you need to ensure that mixing `iterator` and `const_iterator` compiles; if there's no inheritance relationship, you probably need to change how the operators are templated, yes. In vNext, I solved this by templating on `bool _Constant` (instead of encoding `const` into the element type), which allowed me to define operators that were templated on varying constness, but didn't accept wildly differing types. For your operators, you may need to constrain on `is_same_v` after `remove_cv_t`.

Reply:
> `iterator` should be convertible to `const_iterator` but not vice versa
👍 

> Also do we need the non-member friend overloads of those oprators anyway?

I'm happy with the operators as they are now.

Reply:
I am unsure whether `remove_cv_t` is really correct regarding stuff like `span<volatile int>`

Reply:
@StephanTLavavej is there anything still left open here?

Reply:
I believe this is resolved as I see a constrained converting constructor in `_Span_iterator` now, thanks.